vm/src/frame.rs

```diff
@@ -30,16 +30,121 @@ use crate::vm::VirtualMachine;
  * When a name is looked up, it is check in its scope.
  */
 #[derive(Debug)]
+struct RcListNode<T> {
+    elem: T,
+    next: Option<Rc<RcListNode<T>>>,
+}
+
+#[derive(Debug, Clone)]
+struct RcList<T> {
+    head: Option<Rc<RcListNode<T>>>,
+}
+
+struct Iter<'a, T: 'a> {
+    next: Option<&'a RcListNode<T>>,
+}
+
+impl<T> RcList<T> {
+    pub fn new() -> Self {
+        RcList { head: None }
+    }
+
+    pub fn insert(self, elem: T) -> Self {
+        RcList {
+            head: Some(Rc::new(RcListNode {
+                elem,
+                next: self.head,
+            })),
+        }
+    }
+
+    pub fn iter<'a>(&'a self) -> Iter<'a, T> {
+        Iter {
+            next: self.head.as_ref().map(|node| &**node),
+        }
+    }
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.map(|node| {
+            self.next = node.next.as_ref().map(|node| &**node);
+            &node.elem
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
 pub struct Scope {
-    pub locals: PyObjectRef, // Variables
-    // TODO: pub locals: RefCell<PyAttributes>,         // Variables
-    pub parent: Option<Rc<Scope>>, // Parent scope
+    locals: RcList<PyObjectRef>,
+    pub globals: PyObjectRef,
 }
-pub type ScopeRef = Rc<Scope>;
 
 impl Scope {
-    pub fn new(locals: PyObjectRef, parent: Option<ScopeRef>) -> ScopeRef {
-        Rc::new(Scope { locals, parent })
+    pub fn new(locals: Option<PyObjectRef>, globals: PyObjectRef) -> Scope {
+        let locals = match locals {
+            Some(dict) => RcList::new().insert(dict),
+            None => RcList::new(),
+        };
+        Scope { locals, globals }
+    }
+
+    pub fn get_locals(&self) -> PyObjectRef {
+        match self.locals.iter().next() {
+            Some(dict) => dict.clone(),
+            None => self.globals.clone(),
+        }
+    }
+
+    pub fn get_only_locals(&self) -> Option<PyObjectRef> {
+        match self.locals.iter().next() {
+            Some(dict) => Some(dict.clone()),
+            None => None,
+        }
+    }
+
+    pub fn child_scope_with_locals(&self, locals: PyObjectRef) -> Scope {
+        Scope {
+            locals: self.locals.clone().insert(locals),
+            globals: self.globals.clone(),
+        }
+    }
+
+    pub fn child_scope(&self, ctx: &PyContext) -> Scope {
+        self.child_scope_with_locals(ctx.new_dict())
+    }
+}
+
+// TODO: Merge with existing Attribute protocol.
+pub trait AttributeProtocol2 {
```

Comment:
Idea:
```rust
pub trait ScopeProtocol {
    lookup_name(&self, vm, name) -> value;
    define_name(&self, vm, name, value);
    delete_name(&self, vm, name);
}
```

Reply:
Yes. I think you're probably right.

Or maybe NameProtocol to reflect the exception that is raised? I was picturing these as very similar because it works sort of like attributes on classes, and we could give them a common interface - but I don't think there's any reason it has to be the same.

Reply:
I think we should implement this seperate protocol for clarity. Do you wish to do this in this PR?

Reply:
I'll fix this up in this PR. (Maybe not today though.)

I'm still kind of mixed on this. Traits should be to abstract common interfaces from different types. (Which is why I was looking at the interface similarity between attributes and names.) If we're claiming the that it's not the same interface, then should these just be methods directly on Scope itself? (I can't imagine what other types are going to implement ScopeProtocol - the name suggests it would be an odd this to do which is why I suggest NameProtocol.)

I think I'll either:

* Rename AttributeProtocol2 to NameProtocol.
* Move the methods to Scope directly.

Let me know if you have a strong preference (or strongly prefer the name ScopeProtocol). It's easy to change between these, so I guess it doesn't matter.

At this stage I'll probably go with NameProtocol.

Reply:
No strong preference, but I agree with you that no other type will probably implement the name protocol. But then again, you never know. On the other hand, I think it is not good to create many traits which are only used by one type. I would go for NameProtocol only implemented on Scope.

Reply:
Another idea for a name: SymbolProtocol instead of NameProtocol. No strong preference, just another idea!

Reply:
We don't use the word symbol elsewhere though. I think I'll match the method names to the instruction names.